### PR TITLE
URLRouting tweaks

### PR DESCRIPTION
**Pull request by stephencelis**
A few refinements for the experimental routing library.

### CODE REVIEW

File: Sources/Parsing/Parser.swift [hunk 1]
```diff
@@ -59,6 +59,7 @@ extension Parser {
   ///
   /// - Parameter input: A nebulous piece of data to be parsed.
   /// - Returns: A more well-structured value parsed from the given input.
+  @_disfavoredOverload
   @inlinable
   public func parse(_ input: Input) rethrows -> Output {
```
Comment (stephencelis) [inline Sources/Parsing/Parser.swift:64]: This is an easy fallback to match on, but should be disfavored for generic contexts. For example where we should automatically insert `End`s into collection parsers.

File: Sources/Parsing/ParsingError.swift [hunk 2]
```diff
@@ -458,6 +460,17 @@ extension ParsingError.Context {
...
+    case let (lhsInput as Slice<[Substring]>, rhsInput as Substring):
+      return lhsInput.first.map {
+        $0.base != rhsInput.base
+        ? false
+        : $0.startIndex > rhsInput.startIndex
+      }
+      ?? false
+
+    case (is Substring, is Slice<[Substring]>):
+      return !(rhs > lhs)
```
Comment (stephencelis) [inline Sources/Parsing/ParsingError.swift:472]: This properly sorts errors that thread between array slices of substrings and the underlying substrings.

File: Sources/Parsing/ParsingError.swift [hunk 3]
```diff
@@ -486,6 +499,11 @@ private func normalize(_ input: Any) -> Any {
   case let input as Slice<[Substring]>:
     return input.endIndex == input.base.endIndex ? input[..<input.startIndex] : input
 
+  case let input as ArraySlice<Substring>:
+    let base = unsafeBitCast(input._owner!, to: Array<Substring>.self)
```
Comment (stephencelis) [inline Sources/Parsing/ParsingError.swift:503]:
```suggestion
    // NB: This is probably safe to do https://twitter.com/stephentyrone/status/1486507359237251072?s=21&t=zRHJdoSBgCdWxmTzljvZyg
    let base = unsafeBitCast(input._owner!, to: Array<Substring>.self)
```

File: Sources/_URLRouting/PathEnd.swift [hunk 4]
```diff
@@ -4,11 +4,13 @@ public struct PathEnd: ParserPrinter {
 
   @inlinable
   public func parse(_ input: inout URLRequestData) throws {
-    try End().parse(&input.path)
+    guard var first = input.path.first else { return }
+    try End().parse(&first)
```
Comment (stephencelis) [inline Sources/_URLRouting/PathEnd.swift:8]: This pushes the error location into the substring, which is easier to grok.

File: Sources/_URLRouting/URLRequestData.swift [hunk 5]
```diff
@@ -68,17 +68,17 @@ public struct URLRequestData: Equatable, _EmptyInitializable {
     password: String? = nil,
     host: String? = nil,
     port: Int? = nil,
-    path: ArraySlice<Substring> = [],
+    path: String = "",
```
Comment (stephencelis) [inline Sources/_URLRouting/URLRequestData.swift:71]: Updated the `URLRequestData` "designated" initializer to take non-sub-sequences.

File: Tests/_URLRoutingTests/URLRoutingTests.swift [hunk 6]
```diff
@@ -12,6 +12,21 @@ class URLRoutingTests: XCTestCase {
...
 
+  func testPath() {
+    XCTAssertEqual(123, try Path { Int.parser() }.parse(URLRequestData(path: "/123")))
+    XCTAssertThrowsError(try Path { Int.parser() }.parse(URLRequestData(path: "/123-foo"))) { error in
+      XCTAssertEqual(
+        """
+        error: unexpected input
+         --> input:1:5
+        1 | /123-foo
+          |     ^ expected end of input
```
Comment (stephencelis) [inline Tests/_URLRoutingTests/URLRoutingTests.swift:23]: Before the `@_disfavoredOverload` fix, this succeeded by favoring the `parse` overload that doesn't check for the `End`.

File: Sources/_URLRouting/Parsing/ParserPrinter.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+#if canImport(FoundationNetworking)
+  import FoundationNetworking
+#endif
+
+extension ParserPrinter where Input == URLRequestData {
+  @inlinable
+  public func request(for route: Output) -> URLRequest? {
+    try? URLRequest(data: self.print(route))
+  }
```
Comment (mbrandonw) [inline Sources/_URLRouting/Parsing/ParserPrinter.swift:11]:
```suggestion
  public func request(for route: Output) throws -> URLRequest {
    try URLRequest(data: self.print(route))
  }
```

**Review by mbrandonw — APPROVED** [review]

**Pull request merged**